docs: ADR-049 state sync hooks (#10976)
cosmos/cosmos-sdk: docs/architecture/adr-049-state-sync-hooks.md

    SnapshotIAVLItem         iavl              = 2 [(gogoproto.customname) = "IAVL"];
    SnapshotExtensionMeta    extension         = 3;
    SnapshotExtensionPayload extension_payload = 4;
  }
}

// SnapshotExtensionMeta contains metadata about an external snapshotter.
// One module may need multiple snapshotters, so each module may have multiple SnapshotExtensionMeta.
message SnapshotExtensionMeta {
  // the name of the ExtensionSnapshotter, and it is registered to snapshotter manager when setting up the application
  string name   = 1;
  // this is used by each ExtensionSnapshotter to decide the format of payloads included in SnapshotExtensionPayload message
  // it is used within the snapshotter/namespace, not global one for all modules
  uint32 format = 2;
}

// SnapshotExtensionPayload contains payloads of an external snapshotter.
message SnapshotExtensionPayload {
  bytes payload = 1;
}
```

The snapshot stream would look like as follows:

```go
// multi-store snapshot
{SnapshotStoreItem | SnapshotIAVLItem, ...}
// extension1 snapshot
SnapshotExtensionMeta
{SnapshotExtensionPayload, ...}
// extension2 snapshot
SnapshotExtensionMeta

> {SnapshotExtensionPayload, ...}

Reviewer: It's alphabetical by name in current implementation.

> {SnapshotExtensionPayload, ...}

Reviewer 2: Sounds good, would like to see some wording for this. i suppose `name` should be unique then?

> {SnapshotExtensionPayload, ...}

Reviewer: > Sounds good, would like to see some wording for this. i suppose `name` should be unique then (ref: [#10976 (comment)](https://github.com/cosmos/cosmos-sdk/pull/10976#discussion_r791285411))?

yes

stream of `SnapshotStoreItem` and `SnapshotIAVLItem`, which means for app modules that maintain their states outside of
the IAVL tree, they can not add their states to the snapshot stream for state-sync.

## Decision

A simple proposal based on our existing implementation is that, we can add two new message types: `SnapshotExtensionMeta` 
and `SnapshotExtensionPayload`, and they are appended to the existing multi-store stream with `SnapshotExtensionMeta` 
acting as a delimiter between extensions.

```proto
// SnapshotItem is an item contained in a rootmulti.Store snapshot.
message SnapshotItem {
  // item is the specific type of snapshot item.
  oneof item {
    SnapshotStoreItem        store             = 1;
    SnapshotIAVLItem         iavl              = 2 [(gogoproto.customname) = "IAVL"];
    SnapshotExtensionMeta    extension         = 3;
    SnapshotExtensionPayload extension_payload = 4;
  }
}

// SnapshotStoreItem contains metadata about a snapshotted store.
message SnapshotStoreItem {
  string name = 1;
}

// SnapshotIAVLItem is an exported IAVL node.
message SnapshotIAVLItem {
  bytes key     = 1;
  bytes value   = 2;
  int64 version = 3;
  int32 height  = 4;
}

> // SnapshotExtensionMeta contains metadata about an external snapshotter.

Reviewer 3: Here there would be one `SnapshotExtensionMeta` for x/wasm module, maybe a second one for Ethermint module, etc?

Reviewer: Yes, one for each extension snapshotter. You can have multiple snapshotters.

Author: > // One module may need multiple snapshotters, so each module may have multiple SnapshotExtensionMeta.
add a comment and hope that can help clarify the question here

message SnapshotExtensionMeta {
  string name   = 1;

>   uint32 format = 2;

Reviewer 3: format is globally unique?
like x/wasm needs one, ethermint another, agoric a 3rd?
Or is this used within one namespace?

Reviewer: It’s used within the snapshotter/namespace.

Reviewer 4: They don't need to globally unique. They need to be unique in the state machine. So if you are 3 different modules that use different types of state sync packets you need formats 1, 2, 3. I don't see why they should be global

Reviewer: yes, they are not global, they can be 1 at the same time.

Reviewer 3: @zmanian and @yihuang you are saying different things.

> Zaki: They need to be unique in the state machine

This was more or less what I meant by globally unique (on the chain). That you cannot have 2 modules on the same chain both using format 1.

> Yi Huang: yes, they are not global, they can be 1 at the same time.

This is what I understood the answer to be, that `(name, format)` must be unique within this state machine. (And if we keep name globally unique, then we only have to consider format in the context of our own module).

Reviewer 3: Maybe you two can discuss this and determine which interpretation is correct (I am inclined to accept Yi Huang's answer as he is working on this code, but Zaki knows details about the state sync abci framework that the rest of us are missing)

Reviewer: To clarify there are two formats at play in the snapshot:
- one is the snapshot format currently used, it decides both the serialization/chunking and rootmulti snapshot formats.
- then it's the extension snapshotter formats, they are only used by each extension snapshotter to decide the format of payloads included in `SnapshotExtensionPayload` message. So they are not related to one another.

Reviewer: > This is what I understood the answer to be, that (name, format) must be unique within this state machine. (And if we keep name globally unique, then we only have to consider format in the context of our own module).

technically speaking we can have duplicated `name`s in the stream, it just means the snapshotter's `Restore` methods will be called multiple times. But I don't think that makes much sense, maybe we should even validate to prevent that situation.

Reviewer 5: This is somewhat confusing. I recommend updating the field comment to be very explicit.

Reviewer 2: Up. It would be good to clarify if `name` is unique, or `(name, format)` is. IMO first one is simpler.

Reviewer: > Up. It would be good to clarify if `name` is unique, or `(name, format)` is. IMO first one is simpler.

`name` is unique in current implementation.

# ADR 049: State Sync Hooks

## Changelog

- Jan 19, 2022: Initial Draft

## Status

Draft, Under Implementation

## Abstract

> This ADR provides hooks for app modules to publish snapshot of additional state(outside of IAVL tree) for state-sync.

Reviewer 5: ```suggestion
This ADR outlines a hooks-based mechanism for application modules to provide additional state (outside of the IAVL tree) to be used during state sync.
```

## Context

> New clients use state-sync to download snapshots of module state from peer nodes. Currently, the snapshot consists of a

Reviewer 5: ```suggestion
New clients use state-sync to download snapshots of module state from peers. Currently, the snapshot consists of a
```

> stream of `SnapshotStoreItem` and `SnapshotIAVLItem`, which means for app modules that maintain their states outside of
> the IAVL tree, they can not add their states to the snapshot stream for state-sync.

Reviewer 5: ```suggestion
stream of `SnapshotStoreItem` and `SnapshotIAVLItem`, which means that application modules that define their state outside of the IAVL tree cannot include their state as part of the state-sync process.
```

> Notification: `Even though the module state data is outside of the tree, for determinism we
> require that the hash of the external data should be posted in the IAVL tree`

Reviewer 5: ```suggestion
Note, Even though the module state data is outside of the tree, for determinism we
require that the hash of the external data should be posted in the IAVL tree.
```

Besides, we provide `Snapshotter` and `ExtensionSnapshotter` interface for modules to implement snapshotters, which will handle both taking 
snapshot and the restoration. Each module could have mutiple snapshotters, and for modules with additional state, they should
implement `ExtensionSnapshotter` as extension snapshotters. When setting up the application, the snapshot `Manager` should call 
`RegisterExtensions([]ExtensionSnapshotter…)` to register all the extension snapshotters.

```proto
// SnapshotItem is an item contained in a rootmulti.Store snapshot.
message SnapshotItem {
  // item is the specific type of snapshot item.
  oneof item {
    SnapshotStoreItem        store             = 1;
    SnapshotIAVLItem         iavl              = 2 [(gogoproto.customname) = "IAVL"];
    SnapshotExtensionMeta    extension         = 3;
    SnapshotExtensionPayload extension_payload = 4;
  }
}

// SnapshotStoreItem contains metadata about a snapshotted store.
message SnapshotStoreItem {
  string name = 1;
}

// SnapshotIAVLItem is an exported IAVL node.
message SnapshotIAVLItem {
  bytes key     = 1;
  bytes value   = 2;
  int64 version = 3;
  int32 height  = 4;

> }

Reviewer 5: If we're not modifying existing types, I don't think we need to include them here for brevity.

Author: ok , removed `SnapshotStoreItem` and `SnapshotIAVLItem`

// SnapshotExtensionMeta contains metadata about an external snapshotter.
// One module may need multiple snapshotters, so each module may have multiple SnapshotExtensionMeta.
message SnapshotExtensionMeta {
  string name   = 1;
  // format is used within the snapshotter/namespace, not global one for all modules
  uint32 format = 2;
}

// SnapshotExtensionPayload contains payloads of an external snapshotter.
message SnapshotExtensionPayload {
  bytes payload = 1;
}
```

> The snapshot stream would look like this:

Reviewer 5: ```suggestion
The snapshot stream would look like as follows:
```

Author: ```
// SnapshotExtensionMeta contains metadata about an external snapshotter.
// One module may need multiple snapshotters, so each module may have multiple SnapshotExtensionMeta.
message SnapshotExtensionMeta {
  // the name of the ExtensionSnapshotter, and it is registered to snapshotter manager when setting up the application
  string name   = 1;
  // this is used by each ExtensionSnapshotter to decide the format of payloads included in SnapshotExtensionPayload message
  // it is used within the snapshotter/namespace, not global one for all modules
  uint32 format = 2;
}
```
add comments to help explain the fields.

```go
// multi-store snapshot
{SnapshotStoreItem | SnapshotIAVLItem, ...}
// extension1 snapshot
SnapshotExtensionMeta
{SnapshotExtensionPayload, ...}
// extension2 snapshot
SnapshotExtensionMeta
{SnapshotExtensionPayload, ...}
```

> Add `extensions` field to snapshot `Manager` for extension snapshotters. `multistore` snapshotter is a special one and it doesn't need a name because it is always placed at the beginning of the binary stream.

Reviewer 5: ```suggestion
We add an `Extensions` field to snapshot `Manager` for extension snapshotters. The `multistore` snapshotter is a special one and it doesn't need a name because it is always placed at the beginning of the binary stream.
```

```go
type Manager struct {
	store      *Store
	multistore types.Snapshotter

> 	extensions map[string]types.ExtensionSnapshotter
>     mtx                sync.Mutex

Reviewer 5: ```suggestion
	extensions map[string]types.ExtensionSnapshotter
	mtx                sync.Mutex
```

	operation          operation
	chRestore          chan<- io.ReadCloser
	chRestoreDone      <-chan restoreDone
	restoreChunkHashes [][]byte
	restoreChunkIndex  uint32
}
```

For extension snapshotters that implement the `ExtensionSnapshotter` interface, their names should be registered to the snapshot `Manager` by 

> calling `RegisterExtensions` when setting up the application. And the snapshotters will handle both taking snapshot and restoration.

Reviewer 5: ```suggestion
calling `RegisterExtensions` when setting up the application. The snapshotters will handle both taking snapshot and restoration.
```

```go
// RegisterExtensions register extension snapshotters to manager
func (m *Manager) RegisterExtensions(extensions ...types.ExtensionSnapshotter) error 
```

Remain `Snapshotter` interface for `multistore` snapshotter. No need for `SnapshotFormat` and `SupportedFormats` as for `multistore` snapshotter

> formats are handled in a higher level.

Reviewer 5: I don't quite follow this. Consider rephrasing this. What exactly are you trying to describe here?

Author: I have removed this part as the `Snapshotter` interface doesn't change

// ExtensionSnapshotter is an extension Snapshotter that is appended to the snapshot stream.
// ExtensionSnapshotter has an unique name and manages it's own internal formats.
type ExtensionSnapshotter interface {
	Snapshotter

	// SnapshotName returns the name of snapshotter, it should be unique in the manager.
	SnapshotName() string

	// SnapshotFormat returns the default format used to take a snapshot.
	SnapshotFormat() uint32

	// SupportedFormats returns a list of formats it can restore from.
	SupportedFormats() []uint32
}
```

## Consequences

As a result of this implementation, we are able to create snapshots of binary chunk stream for the state that we maintain outside of the IAVL Tree, CosmWasm blobs for example. And new clients are able to fetch sanpshots of state for all modules that have implemented the corresponding interface from peer nodes. 

### Backwards Compatibility

This ADR introduces new proto message types, add field for extension snapshotters to snapshot `Manager`, add new function signatures `SnapshotFormat()`, `SupportedFormats()` to `Snapshotter` interface and add new `ExtensionSnapshotter` interface, so this is not backwards compatible.

### Positive

> State maintained outside of IAVL tree like CosmWasm blobs can create snapshots by implementing extension snapshotters, and being fetched by new clients via state-sync.

Reviewer 5: ```suggestion
- State maintained outside of IAVL tree like CosmWasm blobs can create snapshots by implementing extension snapshotters, and being fetched by new clients via state-sync.
```

### Negative

> // Todo

Reviewer 5: If there are no negatives, which is entirely possible, but unlikely. I would either remove the TODO or note them.

Author: currently negative consequences not found.
would appreciate if anyone can inform of me about this

Reviewer 6: lets leave it blank. The only thing I can think of is increased bandwidth but this isn't really an issue

### Neutral

> All modules that maintain state outside of IAVL tree need to implement `ExtensionSnapshotter` and the snapshot `Manager` need to call `RegisterExtensions` when setting up the application.

Reviewer 5: ```suggestion
- All modules that maintain state outside of IAVL tree need to implement `ExtensionSnapshotter` and the snapshot `Manager` need to call `RegisterExtensions` when setting up the application.
```

A simple proposal based on our existing implementation is that, we can add two new message types: `SnapshotExtensionMeta` 
and `SnapshotExtensionPayload`, and they are appended to the existing multi-store stream with `SnapshotExtensionMeta` 
acting as a delimiter between extensions. Even for modules that maintain the data outside of the tree, for determinism we
require that the hash of the external data should be posted in the IAVL tree. As the chunk hashes should be able to ensure 

> data integrity, we don't need a delimiter to mark the end of the snapshot stream.

Reviewer: I think the purpose of post external state hash into the merkle tree is to commit the external state in the app hash/header hash.

Author: so the chunk hashes should be something different from the hash in the merkle tree?

Reviewer: what do you mean by `, we don't need a delimiter to mark the end of the snapshot stream.`?

Author: I add this sentence to clarify the question from @ethanfrey `I do not see some delimiter to mark the end of said stream, which seems like it would be useful to handle errors. Difference between connection break and finalised stream.`

Reviewer: > I add this sentence to clarify the question from @ethanfrey `I do not see some delimiter to mark the end of said stream, which seems like it would be useful to handle errors. Difference between connection break and finalised stream.`

We have chunk hashes to detect a corrupted stream, so there's no problem here I think.

Author: so chunk hashes should be something different from the hash in the merkle tree, right?

Reviewer: > so chunk hashes should be something different from the hash in the merkle tree, right?

yes, it's different. chunk hashes are for data integrity check during transport, merkle tree hash is for on-chain consensus check.

}
```

For extension snapshotters that implement the `NamedSnapshotter` interface, their names should be registered to the snapshot `Manager` by 

> calling `RegisterExtensions` when setting up the application. And the snapshotters will handle both taking snapshot and restoration.

Reviewer 3: Ah, key point here. Thanks

### Backwards Compatibility

> This ADR introduces new proto message types, add field for extension snapshotters to snapshot `Manager`, add new function signatures `SnapshotFormat()`, `SupportedFormats()` to `Snapshotter` interface and add new `ExtensionSnapshotter` interface, so this is not backwards compatible.

Reviewer: It might be worth mentioning that the wire format of the snapshot is backward compatible when there's no `ExtensionSnapshotter` registered.

Author: sure, I think this is significant for application chains that don't have additional state to take snapshots.
So for application chains of this kind, this ADR should be completely backwards compatible?

Reviewer 2: Yes, let's add some wording that the stream is backwards-compatible if no extension is registered.
